### PR TITLE
Pipeline groups added to generalised WIP view

### LINKED ISSUE #1032: DPL-323 Kanban view for all pipelines [C=M, V=3]
**User story**
As a member of a DNAP lab team, I would like to be able to use the equivalent of the Heron 'work in progress' view, but for the pipelines that I work on, so that I can manage my workload and make sure all plates are processed that need to be.

**Who are the primary contacts for this story**
Liz E, Katy

**Acceptance criteria**
- [x] 'In progress' view is made generic, so it can work for any Limber pipeline.
- [x] There are links for all Limber pipelines from the 'Work in progress' dropdown on the home page.

**Related stories**
https://github.com/sanger/limber/issues/1030 will probably result in the query driving the page being updated - doesn't need to clash with this one but be aware of it.

**Additional context**
Liz E said this would be useful for her team, and also that SSRs might find it useful. 
The idea here is to do the minimum possible to extend it to all pipelines, without doing any extra investigation into the details of any specific pipelines. 
Then we can advertise it to relevant customers and gather user feedback, or make further spin-out stories if any obvious problems arise when applying it to other pipelines.

_Technical details_

There is a dropdown on the home page as follows:
Screenshot 2022-03-21 at 15 56 17
Ideally this would list all pipelines, dynamically pulled from config. These buttons pass a parameter through to the page to tell it which pipeline to show. The relevant section of `pipeline_work_in_progress_controller.rb` is here:
```
# TODO: Add 'pipeline_group' or similar to pipeline config ymls, to group related ones together
# So that it can work for all pipelines
heron_pipeline_name_to_configs = {
  'Heron-384' => ['Heron-384 Tailed A V2', 'Heron-384 Tailed B V2'],
  'Heron-96' => ['Heron-96 Tailed A V2', 'Heron-96 Tailed B V2']
}
```
The challenge is that what lab staff think of a 'pipeline' is sometimes represented by more than one key in the pipeline configuration ymls - hence the suggestion in the code snippet above.

### PR DESCRIPTION
**Pull request by harrietc52**
Closes #1032 

Changes proposed in this pull request:
* Added `pipeline_group` for each pipeline
* Can view WIP labware for any pipeline

Todo:
- [x] Build UI from pipelines, unique, not separate file
- [x] Support empty pipeline group
- [x] Deploy current pipeline groups to UAT, and ask Liz to check if the grouping of pipelines is correct 
- [x] code review fixes
- [x] Katy's controller comment
- [x] finish tests
- [x] see if can refactor `retrieve_pipeline_config_for_group`
- [x] update groups from user emails
- [x] Archive bespoke chrome v2 (new story)
- [x] Check Cardinal with Steve Watts

### CODE REVIEW

**Comment by harrietc52** [comment]
# Deployment

`ansible-playbook -vD deploy_limber.yml -i environments/uat -e "github_branch=3.29.0-dpl-323-1" -e "full_deploy=true reload_nginx=false"`

File: config/pipelines/high_throughput_heron_384_tailed.yml [hunk 1]
```diff
@@ -2,6 +2,7 @@
 # This pipeline can be deleted once all "limber_heron_lthr" requests have progressed through it
 # It is being replaced by high_throughput_heron_384_tailed_v2
 Heron-384 Tailed A: # Heron 384-well pipeline specific to PCR 1 plate
+  pipeline_group: Heron-384
```
Comment (KatyTaylor) [inline config/pipelines/high_throughput_heron_384_tailed.yml:5]: I would call this `Heron-384 (legacy)` or something to make clear it's not used.

File: config/pipelines/high_throughput_cardinal.yml [hunk 2]
```diff
@@ -1,13 +1,16 @@
 ---
 # Initial transfer from tube to plate, as entry for one or other of below pipelines. Happens without submissions / requests.
 Cardinal Entry A:
+  pipeline_group: Cardinal A
```
Comment (KatyTaylor) [inline config/pipelines/high_throughput_cardinal.yml:4]: Change `Cardinal A` to `Cardinal`, or `Cardinal library prep`

File: config/pipelines/high_throughput_cardinal.yml [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
   relationships:
     LCA Blood Vac: LCA Blood Array
 Cardinal Entry B:
+  pipeline_group: Cardinal B
```
Comment (KatyTaylor) [inline config/pipelines/high_throughput_cardinal.yml:8]: Change `Cardinal B` to `Cardinal cell banking`

File: lib/tasks/config.rake [hunk 4]
```diff
@@ -82,6 +82,8 @@ namespace :config do
         end
 
         configuration[:submission_templates] = submission_templates
+
+        # configuration[:pipeline_groups] = pipeline_groups
```
Comment (KatyTaylor) [inline lib/tasks/config.rake:86]: Is this still needed?

File: spec/factories/pipeline_factories.rb [hunk 5]
```diff
@@ -13,7 +13,8 @@
       pipeline_list { Settings.pipelines }
     end
 
-    sequence(:name) { |i| "Pipleine #{i}" }
+    sequence(:name) { |i| "Pipeline #{i}" }
```
Comment (KatyTaylor) [inline spec/factories/pipeline_factories.rb:16]: :)

**Review by KatyTaylor — CHANGES_REQUESTED** [review]
Hey, I think it all probably works well, just had a few suggestions around code readability etc.

I think it would be also good to have some code coverage of the scenario where the pipeline group is missing from the config, and it uses the pipeline name instead.

File: app/models/pipeline.rb [hunk 6]
```diff
@@ -8,6 +8,9 @@
 class Pipeline
   include ActiveModel::Model
 
+  # The group the pipeline belongs to
+  attr_accessor :pipeline_group
```
[the remark this answers is not in the document]
Comment (harrietc52) [reply app/models/pipeline.rb:12]: thats much nicer, thank you

File: config/pipelines/bespoke_aggregation.yml [hunk 7]
```diff
@@ -1,5 +1,6 @@
 ---
 Bespoke Aggregation:
+  pipeline_group: Bespoke Aggregation
```
[the remark this answers is not in the document]
Comment (harrietc52) [reply config/pipelines/bespoke_aggregation.yml]: Sure, I've removed explicit `pipeline_group` for:
Bespoke Aggregation, GBS, Custom Pool, Duplex Seq, RNA Poly A, RNA Poly A Globin, RNA Ribo, RNA Ribo Globin

File: config/pipelines/high_throughput_cardinal.yml (hunk 2 above)
Comment (harrietc52) [reply config/pipelines/high_throughput_cardinal.yml]: Gone with `Cardinal library prep`

File: app/models/pipeline.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (harrietc52) [reply app/models/pipeline.rb]: Seems happy if `pipeline_group` is defined as an `attr_writer`

File: app/controllers/pipeline_work_in_progress_controller.rb [hunk 8]
```diff
@@ -4,15 +4,12 @@
...
 
-    # TODO: Add 'pipeline_group' or similar to pipeline config ymls, to group related ones together
-    # So that it can work for all pipelines
-    heron_pipeline_name_to_configs = {
-      'Heron-384' => ['Heron-384 Tailed A V2', 'Heron-384 Tailed B V2'],
-      'Heron-96' => ['Heron-96 Tailed A V2', 'Heron-96 Tailed B V2']
-    }
-    @ordered_purpose_list = Settings.pipelines.combine_and_order_pipelines(heron_pipeline_name_to_configs[@pipeline])
+    # Group related pipelines together
+    pipeline_groups = Settings.pipelines.build_pipeline_groups(@pipeline)
```
[the remark this answers is not in the document]
Comment (harrietc52) [reply app/controllers/pipeline_work_in_progress_controller.rb:10]: Yep, totally makes sense! And not being fussy at all - both good points that I should have picked up on already 👍

File: app/models/pipeline.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (KatyTaylor) [reply app/models/pipeline.rb]: yep my mistake, I had 2 getters in my example but no setter

**Pull request merged**